MultiThreadSchedulerLocking support #178

Satar07 opened the pull request:
### Description

This PR implements a new Protocol Extension IDET: `MultiThreadSchedulerLocking`.

Closes #178

#### Changes

- Introduced `MultiThreadSchedulerLocking` IDET under `MultiThreadResume`.
- Updated `do_vcont_multi_thread` to track the presence of wildcard/default continue actions.
- If no wildcard continue is found (indicating GDB expects scheduler locking), `gdbstub` now invokes `set_resume_action_scheduler_lock`.
- If the target does not implement this IDET but GDB requests locking, `gdbstub` returns a fatal `PacketUnexpected` error to avoid silent race conditions (as discussed in issue #178).
- Updated the `armv4t_multicore` example to demonstrate a working scheduler-locking implementation.

### API Stability

- \[x\] This PR does not require a breaking API change.

### Checklist

- Documentation
	- \[x\] Ensured any public-facing `rustdoc` formatting looks good
- Validation
	- \[x\] Included output of running `examples/armv4t_multicore`
	- \[x\] Included output of running `./example_no_std/check_size.sh`
- *If implementing a new protocol extension IDET*
	- \[x\] Included a basic sample implementation in `examples/armv4t_multicore`
	- \[x\] IDET can be optimized out (confirmed via `./example_no_std/check_size.sh`)

---

### Validation

#### 1\. Functional Validation (`armv4t_multicore`)

Tested with GDB 15.0. Verified that when `set scheduler-locking on` is used, only the selected thread steps, while the other remain frozen at its last PC.

<details> <summary>GDB Session Output</summary>

Code snippet

```
0x55550000 in ?? ()
(gdb) info threads
  Id   Target Id            Frame
* 1    Thread 1.1 (CPU Cpu) 0x55550000 in ?? ()
  2    Thread 1.2 (CPU Cop) 0x55550000 in ?? ()
(gdb) b*0x55550000+0x30
Breakpoint 1 at 0x55550030
(gdb) c
Continuing.

Thread 1 hit Breakpoint 1, 0x55550030 in ?? ()
(gdb) info threads
  Id   Target Id            Frame
* 1    Thread 1.1 (CPU Cpu) 0x55550030 in ?? ()
  2    Thread 1.2 (CPU Cop) 0x55550044 in ?? ()
(gdb) set scheduler-locking on
(gdb) x/10i $pc
=> 0x55550030:  ldr     r3, [r11, #-16]
   0x55550034:  cmp     r3, #0
   0x55550038:  beq     0x55550030
   0x5555003c:  ldr     r3, [r11, #-8]
   0x55550040:  b       0x55550080
   0x55550044:  mov     r3, #0
   0x55550048:  str     r3, [r11, #-12]
   0x5555004c:  b       0x55550068
   0x55550050:  ldr     r3, [r11, #-8]
   0x55550054:  add     r3, r3, #1
(gdb) b*0x55550038
Breakpoint 3 at 0x55550038
(gdb) c
Continuing.

Thread 1 hit Breakpoint 3, 0x55550038 in ?? ()
(gdb) info threads
  Id   Target Id            Frame
* 1    Thread 1.1 (CPU Cpu) 0x55550038 in ?? ()
  2    Thread 1.2 (CPU Cop) 0x55550044 in ?? ()
```

</details>

#### 2\. Binary Size Check (`./example_no_std/check_size.sh`)

| Section | master | fix/resume\_multi | Diff |
| --- | --- | --- | --- |
| .text | 14128 | 14156 | **+28 bytes** |
| Total | 18586 | 18614 | **+28 bytes** |

<details> <summary>check_size Output</summary>

before:

```
File  .text    Size          Crate Name
2.1%  62.1%  8.6KiB      [Unknown] main
0.2%   5.4%    764B        gdbstub gdbstub::stub::state_machine::GdbStubStateMachineInner<gdbstub::stub::state_machine:...
0.1%   2.4%    336B        gdbstub gdbstub::protocol::commands::breakpoint::BasicBreakpoint::from_slice
0.1%   2.3%    320B        gdbstub <gdbstub::protocol::common::thread_id::ThreadId as core::convert::TryFrom<&[u8]>>::t...
0.1%   2.2%    308B        gdbstub gdbstub::protocol::common::hex::decode_hex_buf
0.1%   2.0%    284B        gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::write
0.1%   1.7%    236B        gdbstub gdbstub::stub::core_impl::resume::<impl gdbstub::stub::core_impl::GdbStubImpl<T,C>>:...
0.1%   1.6%    232B        gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::write_specific_thread_id
0.1%   1.5%    216B           core core::iter::traits::iterator::Iterator::nth
0.1%   1.5%    216B           core core::iter::traits::iterator::Iterator::nth
0.0%   1.1%    160B        gdbstub gdbstub::protocol::common::hex::decode_hex
0.0%   1.1%    156B        gdbstub gdbstub::protocol::common::hex::decode_hex
0.0%   1.1%    156B        gdbstub gdbstub::protocol::common::hex::decode_hex
0.0%   1.1%    156B        gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::write_num
0.0%   1.1%    152B        gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::inner_write
0.0%   1.0%    144B        gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::write_num
0.0%   1.0%    140B        gdbstub <gdbstub::protocol::common::thread_id::IdKind as core::convert::TryFrom<&[u8]>>::try...
0.0%   0.9%    132B        gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::flush
0.0%   0.9%    128B        gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::write_hex
0.0%   0.8%    116B gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiTh...
0.0%   0.8%    108B           core <core::iter::adapters::skip::Skip<I> as core::iter::traits::iterator::Iterator>::next
0.0%   0.7%    104B           core <core::slice::iter::SplitMut<T,P> as core::iter::traits::iterator::Iterator>::next
0.0%   0.5%     76B gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiTh...
0.0%   0.5%     76B gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiTh...
0.0%   0.5%     76B gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiTh...
0.0%   0.5%     68B   gdbstub_arch <gdbstub_arch::arm::reg::arm_core::ArmCoreRegs as gdbstub::arch::Registers>::gdb_des...
0.0%   0.4%     52B gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiTh...
0.0%   0.4%     52B gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiTh...
0.0%   0.4%     52B gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiTh...
0.0%   0.4%     52B      [Unknown] _start
0.0%   0.3%     48B  gdbstub_nostd gdbstub_nostd::print_str::print_str
0.0%   0.1%     20B      [Unknown] call_weak_fn
0.0%   0.1%     12B gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::breakpoints::SwBreakpoint>...
3.3% 100.0% 13.8KiB                .text section size, the file size is 413.5KiB
target/release/gdbstub-nostd  :
section               size     addr
.interp                 27      568
.note.gnu.build-id      36      596
.note.ABI-tag           32      632
.gnu.hash               28      664
.dynsym                432      696
.dynstr                198     1128
.gnu.version            36     1326
.gnu.version_r          48     1368
.rela.dyn              192     1416
.rela.plt              312     1608
.init                   24     1920
.plt                   240     1952
.text                14128     2240
.fini                   20    16368
.rodata                670    16388
.eh_frame_hdr          300    17060
.eh_frame             1072    17360
.init_array              8   130384
.fini_array              8   130392
.dynamic               496   130400
.got                   176   130896
.data                    8   131072
.bss                     8   131080
.comment                87        0
Total                18586
```

after:

```
File  .text    Size          Crate Name
2.1%  62.2%  8.6KiB      [Unknown] main
0.2%   5.4%    764B        gdbstub gdbstub::stub::state_machine::GdbStubStateMachineInner<gdbstub::stub::state_machine:...
0.1%   2.4%    336B        gdbstub gdbstub::protocol::commands::breakpoint::BasicBreakpoint::from_slice
0.1%   2.3%    320B        gdbstub <gdbstub::protocol::common::thread_id::ThreadId as core::convert::TryFrom<&[u8]>>::t...
0.1%   2.2%    308B        gdbstub gdbstub::protocol::common::hex::decode_hex_buf
0.1%   2.0%    284B        gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::write
0.1%   1.7%    236B        gdbstub gdbstub::stub::core_impl::resume::<impl gdbstub::stub::core_impl::GdbStubImpl<T,C>>:...
0.1%   1.6%    232B        gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::write_specific_thread_id
0.1%   1.5%    216B           core core::iter::traits::iterator::Iterator::nth
0.1%   1.5%    216B           core core::iter::traits::iterator::Iterator::nth
0.0%   1.1%    160B        gdbstub gdbstub::protocol::common::hex::decode_hex
0.0%   1.1%    156B        gdbstub gdbstub::protocol::common::hex::decode_hex
0.0%   1.1%    156B        gdbstub gdbstub::protocol::common::hex::decode_hex
0.0%   1.1%    156B        gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::write_num
0.0%   1.1%    152B        gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::inner_write
0.0%   1.0%    144B        gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::write_num
0.0%   1.0%    140B        gdbstub <gdbstub::protocol::common::thread_id::IdKind as core::convert::TryFrom<&[u8]>>::try...
0.0%   0.9%    132B        gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::flush
0.0%   0.9%    128B        gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::write_hex
0.0%   0.8%    116B gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiTh...
0.0%   0.8%    108B           core <core::iter::adapters::skip::Skip<I> as core::iter::traits::iterator::Iterator>::next
0.0%   0.7%    104B           core <core::slice::iter::SplitMut<T,P> as core::iter::traits::iterator::Iterator>::next
0.0%   0.5%     76B gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiTh...
0.0%   0.5%     76B gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiTh...
0.0%   0.5%     76B gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiTh...
0.0%   0.5%     68B   gdbstub_arch <gdbstub_arch::arm::reg::arm_core::ArmCoreRegs as gdbstub::arch::Registers>::gdb_des...
0.0%   0.4%     52B gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiTh...
0.0%   0.4%     52B gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiTh...
0.0%   0.4%     52B gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiTh...
0.0%   0.4%     52B      [Unknown] _start
0.0%   0.3%     48B  gdbstub_nostd gdbstub_nostd::print_str::print_str
0.0%   0.1%     20B      [Unknown] call_weak_fn
0.0%   0.1%     12B gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::breakpoints::SwBreakpoint>...
3.3% 100.0% 13.8KiB                .text section size, the file size is 414.2KiB
target/release/gdbstub-nostd  :
section               size     addr
.interp                 27      568
.note.gnu.build-id      36      596
.note.ABI-tag           32      632
.gnu.hash               28      664
.dynsym                432      696
.dynstr                198     1128
.gnu.version            36     1326
.gnu.version_r          48     1368
.rela.dyn              192     1416
.rela.plt              312     1608
.init                   24     1920
.plt                   240     1952
.text                14156     2240
.fini                   20    16396
.rodata                670    16416
.eh_frame_hdr          300    17088
.eh_frame             1072    17392
.init_array              8   130384
.fini_array              8   130392
.dynamic               496   130400
.got                   176   130896
.data                    8   131072
.bss                     8   131080
.comment                87        0
Total                18614
```

</details>

daniel5151 commented on examples/armv4t_multicore/emu.rs:
```diff
@@ -207,17 +213,20 @@ impl Emu {
         // valid, and you should track which specific TID the GDB client requested to be
         // single-stepped, and run them appropriately.
```
would it be safe to remove this old code comment now that you've added the ability to only step a single core at a time?

daniel5151 commented on src/target/ext/base/multithread.rs:
```diff
@@ -290,3 +313,25 @@ pub trait MultiThreadRangeStepping: Target + MultiThreadResume {
+    /// Configure the target to enable "Scheduler Locking" for the upcoming
+    /// resume.
+    ///
+    /// This method is called when the GDB client sends a `vCont` packet that
```
In general, `gdbstub`'s docs try to avoid leaking the nitty-gritty of the underlying protocol to the end user. Interested users can always crack the code open to figure out these details... but these docs are expected to answer high-level "what is this, how is this useful" questions.

Instead - I would remove this paragraph, and maybe include some phrasing that mentions that this behavior is toggled on/off in GDB by using the `set scheduler-locking {on/off}` command.

daniel5151 commented on src/target/ext/base/multithread.rs:
```diff
@@ -290,3 +313,25 @@ pub trait MultiThreadRangeStepping: Target + MultiThreadResume {
+    ///
+    /// When this method is called, the implementation **must ensure that any
+    /// threads not explicitly resumed** via previous
+    /// `set_resume_action_...` calls **remain stopped**.
```
```suggestion
    /// When this method is called, the implementation must ensure that any
    /// threads not explicitly resumed via previous
    /// `set_resume_action_...` calls **remain stopped**.
```

no need to double emphasis here

daniel5151 reviewed (changes requested):
Great work! Overall design looks good, so after we fix up a few minor doc and error-handling related things, we should be gtg here.

Also, like I mentioned in https://github.com/daniel5151/gdbstub/issues/178#issuecomment-3683126118, it'd be good to double-check that the approach we're talking here lines up with how the GDB client (and maybe `gdbserver`) handle `scheduler-locking`.

I'll see if I can find some time to poke around https://github.com/bminor/binutils-gdb myself, but would def appreciate it if you could poke around a bit as part of this diff.

* * *

Best of luck with exams! I'm also somewhat busy this week (home for the holidays and whatnot), so no need to rush or anything.

daniel5151 commented:
> Also, like I mentioned in [#178 (comment)](https://github.com/daniel5151/gdbstub/issues/178#issuecomment-3683126118), it'd be good to double-check that the approach we're talking here lines up with how the GDB client (and maybe `gdbserver`) handle `scheduler-locking`.

I nerd sniped myself, so I did some digging here lol.

https://github.com/bminor/binutils-gdb/blob/b0196f0/gdb/remote.c#L960-L961

It _looks_ like GDB unconditionally assumes that the remote target supports scheduler-locking, which seems... not ideal. Interestingly enough - I chased the `git blame` chain to see when this default was introduced, and it looks like it stretches waaaaaaaaaay back into the past (I'm talking before 1999, and even before the initial commit that migrated GDB over to git). Check it out: https://github.com/bminor/binutils-gdb/blob/c906108c21474dfb4ed285bcc0ac6fe02cd400cc/gdb/remote.c#L3389

So, if you want some bonus points - you might want to consider opening an issue (or even sending in a PR?) to GDB that points out this behavior, and that it might be more user-friendly if GDB negotiated support for scheduler-locking on remote targets via a `qSupported` feature.

Satar07 committed:
implement MultiThreadSchedulerLocking support

- Protocol: Add `MultiThreadSchedulerLocking` IDET to track and enforce
  GDB's scheduler-locking mode during `vCont` handling.
- Error Handling: Introduce `MissingMultiThreadSchedulerLocking`
  internal error to provide clear feedback when the IDET is missing.
- Documentation: Refactor IDET docs to focus on high-level GDB user
  behavior (`set scheduler-locking`) instead of RSP packet details.
- Examples: Update `armv4t_multicore` to support core freezing via
  `ExecMode::Stop`, providing a reference implementation for lock-step
  multi-core targets.
- Examples: Revert unnecessary cycle stall changes in the emulator.

Closes #178

Satar07 commented on src/stub/core_impl/resume.rs:
```diff
@@ -251,6 +259,18 @@ impl<T: Target, C: Connection> GdbStubImpl<T, C> {
+                error!(
+                    "GDB requested Scheduler Locking (vCont packet missing wildcard action), but the Target does not implement the `MultiThreadSchedulerLocking` IDET."
+                );
+                return Err(Error::PacketUnexpected);
```
Fixed! Added the InternalError variant. Thanks for the tip, learned a lot about the error handling here

marc-hb commented:
I don't whether this is relevant, but sharing just in case:

```
Reading symbols from ../../../target/thumbv7em-none-eabihf/debug/examples/init...
(gdb) set non-stop on
(gdb) target remote :1337
Remote debugging using :1337
❌️ Non-stop mode requested, but remote does not support non-stop
(gdb) 
```
On the gdbstub side:
```
     GDB stub listening at 127.0.0.1:1337

  < waits until target remote connects>

During the execution of GDB an error was encountered:
failed to fill whole buffer
        Done processing config profile default
```
